org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/AST.java

```diff
@@ -1243,7 +1243,7 @@ private AST(int level, boolean previewEnabled) {
 	 * @see JavaCore#getDefaultOptions()
 	 */
 	public AST(Map options) {
-		this(apiLevelMap.get(options.get(JavaCore.COMPILER_SOURCE)),
+		this(apiLevelMap.getOrDefault(JavaCore.COMPILER_SOURCE, JLS_INTERNAL_Latest),
```

Comment:
Should it not better default to the _lowest_ known value? I can only think about that values become "illegal" after a while when older java versions are dropped, and while development time it would be better to get possibly an error for an unsupported java feature when specify a too high value.

Reply:
> Should it not better default to the lowest known value? 

If using the lowest version, then every more modern structure will cause exceptions later and result in a non usable AST/editor. And as the initial report is about using a too new version, going to latest is the least breaking default.

Reply:
> If using the lowest version, then every more modern structure will cause exceptions later and result in a non usable AST/editor

I think setting a "too high" version manually (or by code) is an error and one not really gets "unusable" AST/editor but just complaining about unusupported things and one then might check this.

>  And as the initial report is about using a too new version, going to latest is the least breaking default.

The report mainly complains about a NPE so I'm not sure what was the behavior before, but if it set e.g. java version =  99 and get no errors my expectation would be that this value is used. but maybe @stephan-herrmann can tell what is the actual expectation from compiler point of view.

Reply:
> I think setting a "too high" version manually (or by code) is an error

Not necessarily. Opening a recent project with source version set in metadata in an older Eclipse version leads to this state and is not so much of an error.

Reply:
> > I think setting a "too high" version manually (or by code) is an error
> 
> Not necessarily. Opening a recent project with source version set in metadata in an older Eclipse version leads to this state and is not so much of an error.

But if my "recent project" requires Java 24 but my "older Eclipse" only supports Java 21, then it does not help much if if get Java 21 as the project settings that complains about Java 24 features.

So I still think this is an error open a project requiring a higher java version in an Eclipse not supporting that.

Reply:
To be honest, I don't really care about the chosen solution in the end as it's not a problem that is in my current scope of interest, so please submit an alternative PR withou your proposal and if it's better, let's just merge it.

Reply:
@mickaelistria  @laeubi  I can see where the disagreement might come from: creating an `AST` for an unsupported version is "unsupported" - so there may not be a perfect solution.

For the original problem #3929  it may still be better to avoid (needlessly) using this API for #2700 as mentioned in https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3929#issuecomment-2805147096

Reply:
BTW:
> it's not a problem that is in my current scope of interest

it may well enter your scope of interest if #3929 is fixed by reverting #2700 which triggered the NPE :)